concordium-sdk/src/main/java/com/concordium/sdk/transactions/smartcontracts/WasmModule.java

```diff
@@ -107,4 +121,55 @@ public byte[] getBytes() {
         return buffer.array();
     }
 
+    /**
+     * Retrieve the {@link Schema} corresponding to the contract, if embedded.
+     * Behaviour is not specified if the bytes of {@link WasmModule} do not represent a valid concordium wasm module.
+     * @return {@link Optional} containing the {@link Schema} if found, empty otherwise.
+     */
+    public Optional<Schema> getSchema() {
+        val moduleSourceBytes = source.getBytes().clone();
+        val buffer = ByteBuffer.wrap(moduleSourceBytes);
+
+        // Skip 4 byte length of WasmModuleSource (UInt32.BYTES) + 4 byte magic number + 4 byte WASM version
+        buffer.position(buffer.position() + 12);
+
+        while (buffer.hasRemaining()) {
+            // A section is a 1 byte id followed by the length of the section as a LEB128U encoded u32.
+            byte id = buffer.get();
+            int remainingSectionLength = LEB128U.decode(buffer, LEB128U.U32_BYTES).intValue();
+
+            // Custom sections have id 0 so all other ids are skipped.
+            if (id != 0) {
+                buffer.position(buffer.position() + remainingSectionLength);
+                continue;
+            }
+
+            // Custom sections have a name encoded as a vector i.e. a length followed by the actual bytes
+            int beforeName = buffer.position();
+            int nameLength = LEB128U.decode(buffer, LEB128U.U32_BYTES).intValue();
+            int nameLengthBytes = buffer.position() - beforeName;
+
+            byte[] nameBytes = new byte[nameLength];
+            buffer.get(nameBytes);
+
+            String name = Strings.fromByteArray(nameBytes);
+
+            // We've incremented the buffer by reading the length of the name and the name.
+            remainingSectionLength = remainingSectionLength - nameLengthBytes - nameLength;
+
+            if (name.equals("concordium-schema")) {
```

Comment:
In principle there are also `concordium-schema-v1` and `concordium-schema-v2` , see https://docs.rs/axum/latest/axum/struct.Router.html#method.fallback_service https://github.com/Concordium/concordium-base/blob/87a752f185869ea567387baac384e0e5bc997fd9/smart-contracts/wasm-chain-integration/src/utils.rs#L583

which would be great to support since the Schema class does as well as far as I understand.

The difference is that the `concordium-schema` section contains the schema that is versioned inside.

Whereas the new (old,legacy) `-v1` and `-v2` sections, the contents is the unversioned schema and the version is in the name (v0 and v1).